README.md
# proper-lockfile

[![NPM version][npm-image]][npm-url] [![Downloads][downloads-image]][npm-url] [![Build Status][travis-image]][travis-url] [![Coverage Status][codecov-image]][codecov-url] [![Dependency status][david-dm-image]][david-dm-url] [![Dev Dependency status][david-dm-dev-image]][david-dm-dev-url]

[npm-url]:https://npmjs.org/package/proper-lockfile
[downloads-image]:http://img.shields.io/npm/dm/proper-lockfile.svg
[npm-image]:http://img.shields.io/npm/v/proper-lockfile.svg
[travis-url]:https://travis-ci.org/moxystudio/node-proper-lockfile
[travis-image]:http://img.shields.io/travis/moxystudio/node-proper-lockfile/master.svg
[codecov-url]:https://codecov.io/gh/moxystudio/node-proper-lockfile
[codecov-image]:https://img.shields.io/codecov/c/github/moxystudio/node-proper-lockfile/master.svg
[david-dm-url]:https://david-dm.org/moxystudio/node-proper-lockfile
[david-dm-image]:https://img.shields.io/david/moxystudio/node-proper-lockfile.svg
[david-dm-dev-url]:https://david-dm.org/moxystudio/node-proper-lockfile?type=dev
[david-dm-dev-image]:https://img.shields.io/david/dev/moxystudio/node-proper-lockfile.svg

An inter-process and inter-machine lockfile utility that works on a local or network file system.


## Installation

`$ npm install proper-lockfile`


## Design

There are various ways to achieve [file locking](http://en.wikipedia.org/wiki/File_locking).

This library utilizes the `mkdir` strategy which works atomically on any kind of file system, even network based ones.
The lockfile path is based on the file path you are trying to lock by suffixing it with `.lock`.

When a lock is successfully acquired, the lockfile's `mtime` (modified time) is periodically updated to prevent staleness. This allows to effectively check if a lock is stale by checking its `mtime` against a stale threshold. If the update of the mtime fails several times, the lock might be compromised. The `mtime` is [supported](http://en.wikipedia.org/wiki/Comparison_of_file_systems) in almost every `filesystem`.


### Comparison

This library is similar to [lockfile](https://github.com/isaacs/lockfile) but the later has some drawbacks:

- It relies on `open` with `O_EXCL` flag which has problems in network file systems. `proper-lockfile` uses `mkdir` which doesn't have this issue.

> O_EXCL is broken on NFS file systems; programs which rely on it for performing locking tasks will contain a race condition.

- The lockfile staleness check is done via `ctime` (creation time) which is unsuitable for long running processes. `proper-lockfile` constantly updates lockfiles `mtime` to do proper staleness check.

- It does not check if the lockfile was compromised which can led to undesirable situations. `proper-lockfile` checks the lockfile when updating the `mtime`.

- It has a default value of `0` for the stale option which isn't good because any crash or process kill that the package can't handle gracefully will leave the lock active forever.


### Compromised

`proper-lockfile` does not detect cases in which:

- A `lockfile` is manually removed and someone else acquires the lock right after
- Different `stale`/`update` values are being used for the same file, possibly causing two locks to be acquired on the same file

`proper-lockfile` detects cases in which:

- Updates to the `lockfile` fail
- Updates take longer than expected, possibly causing the lock to became stale for a certain amount of time


As you see, the first two are a consequence of bad usage. Technically, it was possible to detect the first two but it would introduce complexity and eventual race conditions.


## Usage

### .lock(file, [options])

Tries to acquire a lock on `file` or rejects the promise by throwing on error.

If the lock succeeds, a `release` function is provided that should be called when you want to release the lock. The `release` function also rejects the promise by throwing on error (e.g. when the lock was already compromised).

Available options:

- `stale`: Duration in milliseconds in which the lock is considered stale, defaults to `10000` (minimum value is `5000`)
- `update`: The interval in milliseconds in which the lockfile's `mtime` will be updated, defaults to `stale/2` (minimum value is `1000`, maximum value is `stale/2`)
- `retries`: The number of retries or a [retry](https://www.npmjs.org/package/retry) options object, defaults to `0`
- `realpath`: Resolve symlinks using realpath, defaults to `true` (note that if `true`, the `file` must exist previously)
- `fs`: A custom fs to use, defaults to `graceful-fs`
- `onCompromised`: Called if the lock gets compromised, defaults to a function that simply throws the error which will probably cause the process to die
- `lockfilePath`: Custom lockfile path. e.g.: If you want to lock a directory and create the lock file inside it, you can pass `file` as `<dir path>` and `options.lockfilePath` as `<dir path>/dir.lock`


```js
const lockfile = require('proper-lockfile');

lockfile.lock('some/file')
.then((release) => {
    // Do something while the file is locked

    // Call the provided release function when you're done,
    // which will also return a promise
    return release();
})
.catch((e) => {
    // either lock could not be acquired
    // or releasing it failed
    console.error(e)
});

// Alternatively, you may use lockfile('some/file') directly.
```


### .unlock(file, [options])

Releases a previously acquired lock on `file` or rejects the promise by throwing on error.

Whenever possible you should use the `release` function instead (as exemplified above). Still there are cases in which its hard to keep a reference to it around code. In those cases `unlock()` might be handy.

Available options:

- `realpath`: Resolve symlinks using realpath, defaults to `true` (note that if `true`, the `file` must exist previously)
- `fs`: A custom fs to use, defaults to `graceful-fs`
- `lockfilePath`: Custom lockfile path. e.g.: If you want to lock a directory and create the lock file inside it, you can pass `file` as `<dir path>` and `options.lockfilePath` as `<dir path>/dir.lock`


```js
const lockfile = require('proper-lockfile');

lockfile.lock('some/file')
.then(() => {
    // Do something while the file is locked

    // Later..
    return lockfile.unlock('some/file');
});
```

### .check(file, [options])

Check if the file is locked and its lockfile is not stale, rejects the promise by throwing on error.

Available options:

- `stale`: Duration in milliseconds in which the lock is considered stale, defaults to `10000` (minimum value is `5000`)
- `realpath`: Resolve symlinks using realpath, defaults to `true` (note that if `true`, the `file` must exist previously)
- `fs`: A custom fs to use, defaults to `graceful-fs`
- `lockfilePath`: Custom lockfile path. e.g.: If you want to lock a directory and create the lock file inside it, you can pass `file` as `<dir path>` and `options.lockfilePath` as `<dir path>/dir.lock`


```js
const lockfile = require('proper-lockfile');

lockfile.check('some/file')
.then((isLocked) => {
    // isLocked will be true if 'some/file' is locked, false otherwise
});
```

### .lockSync(file, [options])

Sync version of `.lock()`.   
Returns the `release` function or throws on error.

### .unlockSync(file, [options])

Sync version of `.unlock()`.   
Throws on error.

### .checkSync(file, [options])

Sync version of `.check()`.
Returns a boolean or throws on error.


## Graceful exit

`proper-lockfile` automatically remove locks if the process exists, except if the process is killed with SIGKILL or it crashes due to a VM fatal error (e.g.: out of memory).


## Tests

`$ npm test`   
`$ npm test -- --watch` during development

The test suite is very extensive. There's even a stress test to guarantee exclusiveness of locks.


## License

Released under the [MIT License](http://www.opensource.org/licenses/mit-license.php).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="4.1.0"></a>
# [4.1.0](https://github.com/moxystudio/node-proper-lockfile/compare/v4.0.0...v4.1.0) (2019-03-18)


### Features

* allow second precision in mtime comparison ([#78](https://github.com/moxystudio/node-proper-lockfile/issues/78)) ([b2816a6](https://github.com/moxystudio/node-proper-lockfile/commit/b2816a6))



<a name="4.0.0"></a>
# [4.0.0](https://github.com/moxystudio/node-proper-lockfile/compare/v3.2.0...v4.0.0) (2019-03-12)


### Bug Fixes

* fix typo in error message ([#68](https://github.com/moxystudio/node-proper-lockfile/issues/68)) ([b91cb55](https://github.com/moxystudio/node-proper-lockfile/commit/b91cb55))


### Features

* make staleness check more robust ([#74](https://github.com/moxystudio/node-proper-lockfile/issues/74)) ([9cc0973](https://github.com/moxystudio/node-proper-lockfile/commit/9cc0973)), closes [#71](https://github.com/moxystudio/node-proper-lockfile/issues/71) [/github.com/ipfs/js-ipfs-repo/issues/188#issuecomment-468682971](https://github.com//github.com/ipfs/js-ipfs-repo/issues/188/issues/issuecomment-468682971)


### BREAKING CHANGES

* We were marking the lock as compromised when system went into sleep or if the event loop was busy taking too long to run the internals timers, Now we keep track of the mtime updated by the current process, and if we lose some cycles in the update process but recover and the mtime is still ours we do not mark the lock as compromised.



<a name="3.2.0"></a>
# [3.2.0](https://github.com/moxystudio/node-proper-lockfile/compare/v3.1.0...v3.2.0) (2018-11-19)


### Features

* add lock path option ([#66](https://github.com/moxystudio/node-proper-lockfile/issues/66)) ([32f1b8d](https://github.com/moxystudio/node-proper-lockfile/commit/32f1b8d))



<a name="3.1.0"></a>
# [3.1.0](https://github.com/moxystudio/node-proper-lockfile/compare/v3.0.2...v3.1.0) (2018-11-15)


### Bug Fixes

* **package:** update retry to version 0.12.0 ([#50](https://github.com/moxystudio/node-proper-lockfile/issues/50)) ([d400b98](https://github.com/moxystudio/node-proper-lockfile/commit/d400b98))


### Features

* add signal exit ([#65](https://github.com/moxystudio/node-proper-lockfile/issues/65)) ([f20bc45](https://github.com/moxystudio/node-proper-lockfile/commit/f20bc45))



<a name="3.0.2"></a>
## [3.0.2](https://github.com/moxystudio/node-proper-lockfile/compare/v3.0.1...v3.0.2) (2018-01-30)



<a name="3.0.1"></a>
## [3.0.1](https://github.com/moxystudio/node-proper-lockfile/compare/v3.0.0...v3.0.1) (2018-01-20)


### Bug Fixes

* restore ability to use lockfile() directly ([0ef8fbc](https://github.com/moxystudio/node-proper-lockfile/commit/0ef8fbc))



<a name="3.0.0"></a>
# [3.0.0](https://github.com/moxystudio/node-proper-lockfile/compare/v2.0.1...v3.0.0) (2018-01-20)


### Chores

* update project to latest node lts ([b1d43e5](https://github.com/moxystudio/node-proper-lockfile/commit/b1d43e5))


### BREAKING CHANGES

* remove callback support
* use of node lts language features such as object spread
* compromised function in lock() has been moved to an option


test/misc.test.js
'use strict';

const fs = require('fs');
const execa = require('execa');
const mkdirp = require('mkdirp');
const rimraf = require('rimraf');

const tmpDir = `${__dirname}/tmp`;

beforeAll(() => mkdirp.sync(tmpDir));

afterAll(() => rimraf.sync(tmpDir));

afterEach(() => rimraf.sync(`${tmpDir}/*`));

it('should always use `options.fs` when calling `fs` methods', () => {
    const lockfileContents = fs.readFileSync(`${__dirname}/../lib/lockfile.js`);

    expect(/\s{1,}fs\.[a-z]+/i.test(lockfileContents)).toBe(false);
});

it('should remove open locks if the process crashes', async () => {
    const { stderr } = await execa('node', [`${__dirname}/fixtures/crash.js`], { reject: false });

    expect(stderr).toMatch('intencional crash');
    expect(fs.existsSync(`${tmpDir}/foo.lock`)).toBe(false);
});

it('should not hold the process if it has no more work to do', async () => {
    await execa('node', [`${__dirname}/fixtures/unref.js`]);
});

it('should work on stress conditions', async () => {
    try {
        await execa('node', [`${__dirname}/fixtures/stress.js`]);
    } catch (err) {
        const stdout = err.stdout || '';

        if (process.env.CI) {
            process.stdout.write(stdout);
        } else {
            fs.writeFileSync(`${__dirname}/stress.log`, stdout);
        }

        throw err;
    }
}, 80000);


index.js
'use strict';

const lockfile = require('./lib/lockfile');
const { toPromise, toSync, toSyncOptions } = require('./lib/adapter');

async function lock(file, options) {
    const release = await toPromise(lockfile.lock)(file, options);

    return toPromise(release);
}

function lockSync(file, options) {
    const release = toSync(lockfile.lock)(file, toSyncOptions(options));

    return toSync(release);
}

function unlock(file, options) {
    return toPromise(lockfile.unlock)(file, options);
}

function unlockSync(file, options) {
    return toSync(lockfile.unlock)(file, toSyncOptions(options));
}

function check(file, options) {
    return toPromise(lockfile.check)(file, options);
}

function checkSync(file, options) {
    return toSync(lockfile.check)(file, toSyncOptions(options));
}

module.exports.lock = lock;
module.exports.unlock = unlock;
module.exports.lockSync = lockSync;
module.exports.unlockSync = unlockSync;
module.exports.check = check;
module.exports.checkSync = checkSync;


lib/mtime.js
'use strict';

const fs = require('graceful-fs');

const MTIME = new Date(1552815428391);

function getPrecision(path, options, callback) {
    options = options || {};
    options.fs = options.fs || fs;

    options.fs.stat(path, (err, stat) => {
        if (err) {
            return callback(err);
        }

        // Attempt to update mtime to our known past value
        options.fs.utimes(path, stat.atime, MTIME, (err) => {
            if (err) {
                return callback(err);
            }

            // Read back the mtime and see if it is still ms precision
            options.fs.stat(path, (err, updatedStat) => {
                if (err) {
                    return callback(err);
                }

                // Put back the original mtime, nothing to see here!
                options.fs.utimes(path, stat.atime, stat.mtime, (err) => {
                    if (err) {
                        return callback(err);
                    }

                    if (updatedStat.mtime.getTime() === MTIME.getTime()) {
                        callback(null, 'ms');
                    } else if (updatedStat.mtime.getTime() === (Math.trunc(MTIME.getTime() / 1000) * 1000)) {
                        callback(null, 's');
                    } else {
                        callback(Object.assign(
                            new Error('Unsupported mtime precision'),
                            { code: 'EMTIMEPRECISION' }
                        ));
                    }
                });
            });
        });
    });
}

// Determine if mtimes are equal according to precision
function isEqualMtimes(t0, t1, precision) {
    if (t0.getTime() === t1.getTime()) {
        return true;
    }

    if (precision === 'ms') {
        return false;
    }

    if (precision === 's') {
        return Math.trunc(t0.getTime() / 1000) === Math.trunc(t1.getTime() / 1000);
    }

    throw new Error(`Unknown precision ${precision}`);
}

module.exports.getPrecision = getPrecision;
module.exports.isEqualMtimes = isEqualMtimes;


lib/mtimes.js
'use strict';

const fs = require('graceful-fs');

const MTIME = new Date(1552815428391);

function getPrecision(path, options, callback) {
    options = options || {};
    options.fs = options.fs || fs;

    options.fs.stat(path, (err, stat) => {
        if (err) {
            return callback(err);
        }

        // Attempt to update mtime to our known past value
        options.fs.utimes(path, stat.atime, MTIME, (err) => {
            if (err) {
                return callback(err);
            }

            // Read back the mtime and see if it is still ms precision
            options.fs.stat(path, (err, updatedStat) => {
                if (err) {
                    return callback(err);
                }

                // Put back the original mtime, nothing to see here!
                options.fs.utimes(path, stat.atime, stat.mtime, (err) => {
                    if (err) {
                        return callback(err);
                    }

                    if (updatedStat.mtime.getTime() === MTIME.getTime()) {
                        callback(null, 'ms');
                    } else if (updatedStat.mtime.getTime() === (Math.trunc(MTIME.getTime() / 1000) * 1000)) {
                        callback(null, 's');
                    } else {
                        callback(Object.assign(
                            new Error('Unsupported mtime precision'),
                            { code: 'EMTIMEPRECISION' }
                        ));
                    }
                });
            });
        });
    });
}

// Determine if mtimes are equal according to precision
function isEqual(t0, t1, precision) {
    if (t0.getTime() === t1.getTime()) {
        return true;
    }

    if (precision === 'ms') {
        return false;
    }

    if (precision === 's') {
        return Math.trunc(t0.getTime() / 1000) === Math.trunc(t1.getTime() / 1000);
    }

    throw new Error(`Unknown precision ${precision}`);
}

module.exports.getPrecision = getPrecision;
module.exports.isEqual = isEqual;


lib/syncFs.js
'use strict';

function makeSync(fs, name) {
    var fn = fs[name + 'Sync'];

    return function () {
        var callback = arguments[arguments.length - 1],
            args = Array.prototype.slice.call(arguments, 0, -1),
            ret;

        try {
            ret = fn.apply(fs, args);
            callback(null, ret);
        } catch (err) {
            callback(err);
        }
    };
}

function syncFs(fs) {
    var fns = ['mkdir', 'realpath', 'stat', 'rmdir', 'utimes'],
        obj = {},
        key;

    // Create the sync versions of the methods that we need
    fns.forEach(function (name) {
        obj[name] = makeSync(fs, name);
    });

    // Copy the rest of the functions
    for (key in fs) {
        if (!obj[key]) {
            obj[key] = fs[key];
        }
    }

    return obj;
}

module.exports = syncFs;


lib/mtime-precision.js
'use strict';

const fs = require('graceful-fs');

const MTIME = new Date(1552815428391);

function getPrecision(path, options, callback) {
    options = options || {};
    options.fs = options.fs || fs;

    options.fs.stat(path, (err, stat) => {
        if (err) {
            return callback(err);
        }

        // Attempt to update mtime to our known past value
        options.fs.utimes(path, stat.atime, MTIME, (err) => {
            if (err) {
                return callback(err);
            }

            // Read back the mtime and see if it is still ms precision
            options.fs.stat(path, (err, updatedStat) => {
                if (err) {
                    return callback(err);
                }

                // Put back the original mtime, nothing to see here!
                options.fs.utimes(path, stat.atime, stat.mtime, (err) => {
                    if (err) {
                        return callback(err);
                    }

                    if (updatedStat.mtime.getTime() === MTIME.getTime()) {
                        callback(null, 'ms');
                    } else if (updatedStat.mtime.getTime() === (Math.trunc(MTIME.getTime() / 1000) * 1000)) {
                        callback(null, 's');
                    } else {
                        callback(Object.assign(
                            new Error('Unsupported mtime precision'),
                            { code: 'EMTIMEPRECISION' }
                        ));
                    }
                });
            });
        });
    });
}

module.exports = getPrecision;


lib/lockfile.js
'use strict';

const path = require('path');
const fs = require('graceful-fs');
const retry = require('retry');
const onExit = require('signal-exit');

const locks = {};

function getLockFile(file, options) {
    return options.lockfilePath || `${file}.lock`;
}

function resolveCanonicalPath(file, options, callback) {
    if (!options.realpath) {
        return callback(null, path.resolve(file));
    }

    // Use realpath to resolve symlinks
    // It also resolves relative paths
    options.fs.realpath(file, callback);
}

function acquireLock(file, options, callback) {
    // Use mkdir to create the lockfile (atomic operation)
    options.fs.mkdir(getLockFile(file, options), (err) => {
        // If successful, we are done
        if (!err) {
            return options.fs.stat(getLockFile(file, options), callback);
        }

        // If error is not EEXIST then some other error occurred while locking
        if (err.code !== 'EEXIST') {
            return callback(err);
        }

        // Otherwise, check if lock is stale by analyzing the file mtime
        if (options.stale <= 0) {
            return callback(Object.assign(new Error('Lock file is already being held'), { code: 'ELOCKED', file }));
        }

        options.fs.stat(getLockFile(file, options), (err, stat) => {
            if (err) {
                // Retry if the lockfile has been removed (meanwhile)
                // Skip stale check to avoid recursiveness
                if (err.code === 'ENOENT') {
                    return acquireLock(file, { ...options, stale: 0 }, callback);
                }

                return callback(err);
            }

            if (!isLockStale(stat, options)) {
                return callback(Object.assign(new Error('Lock file is already being hold'), { code: 'ELOCKED', file }));
            }

            // If it's stale, remove it and try again!
            // Skip stale check to avoid recursiveness
            removeLock(file, options, (err) => {
                if (err) {
                    return callback(err);
                }

                acquireLock(file, { ...options, stale: 0 }, callback);
            });
        });
    });
}

function isLockStale(stat, options) {
    return stat.mtime.getTime() < Date.now() - options.stale;
}

function removeLock(file, options, callback) {
    // Remove lockfile, ignoring ENOENT errors
    options.fs.rmdir(getLockFile(file, options), (err) => {
        if (err && err.code !== 'ENOENT') {
            return callback(err);
        }

        callback();
    });
}

function updateLock(file, options) {
    const lock = locks[file];

    // Just for safety, should never happen
    /* istanbul ignore if */
    if (lock.updateTimeout) {
        return;
    }

    lock.updateDelay = lock.updateDelay || options.update;
    lock.updateTimeout = setTimeout(() => {
        lock.updateTimeout = null;

        // Check if mtime is still ours if it is we can still recover from a system sleep or a busy event loop
        options.fs.stat(getLockFile(file, options), (err, stat) => {
            const isOverThreshold = lock.lastUpdate + options.stale < Date.now();

            // If it failed to update the lockfile, keep trying unless
            // the lockfile was deleted or we are over the threshold
            if (err) {
                if (err.code === 'ENOENT' || isOverThreshold) {
                    return setLockAsCompromised(file, lock, Object.assign(err, { code: 'ECOMPROMISED' }));
                }

                lock.updateDelay = 1000;

                return updateLock(file, options);
            }

            const isMtimeOurs = lock.mtimeChecker(lock.mtime, stat.mtime);

            if (!isMtimeOurs) {
                return setLockAsCompromised(
                    file,
                    lock,
                    Object.assign(
                        new Error('Unable to update lock within the stale threshold'),
                        { code: 'ECOMPROMISED' }
                    ));
            }

            const mtime = new Date();

            options.fs.utimes(getLockFile(file, options), mtime, mtime, (err) => {
                const isOverThreshold = lock.lastUpdate + options.stale < Date.now();

                // Ignore if the lock was released
                if (lock.released) {
                    return;
                }

                // If it failed to update the lockfile, keep trying unless
                // the lockfile was deleted or we are over the threshold
                if (err) {
                    if (err.code === 'ENOENT' || isOverThreshold) {
                        return setLockAsCompromised(file, lock, Object.assign(err, { code: 'ECOMPROMISED' }));
                    }

                    lock.updateDelay = 1000;

                    return updateLock(file, options);
                }

                // All ok, keep updating..
                lock.mtime = mtime;
                lock.lastUpdate = Date.now();
                lock.updateDelay = null;
                updateLock(file, options);
            });
        });
    }, lock.updateDelay);

    // Unref the timer so that the nodejs process can exit freely
    // This is safe because all acquired locks will be automatically released
    // on process exit

    // We first check that `lock.updateTimeout.unref` exists because some users
    // may be using this module outside of NodeJS (e.g., in an electron app),
    // and in those cases `setTimeout` return an integer.
    /* istanbul ignore else */
    if (lock.updateTimeout.unref) {
        lock.updateTimeout.unref();
    }
}

function setLockAsCompromised(file, lock, err) {
    // Signal the lock has been released
    lock.released = true;

    // Cancel lock mtime update
    // Just for safety, at this point updateTimeout should be null
    /* istanbul ignore if */
    if (lock.updateTimeout) {
        clearTimeout(lock.updateTimeout);
    }

    if (locks[file] === lock) {
        delete locks[file];
    }

    lock.options.onCompromised(err);
}

// ----------------------------------------------------------

function lock(file, options, callback) {
    /* istanbul ignore next */
    options = {
        stale: 10000,
        update: null,
        realpath: true,
        retries: 0,
        fs,
        onCompromised: (err) => { throw err; },
        ...options,
    };

    options.retries = options.retries || 0;
    options.retries = typeof options.retries === 'number' ? { retries: options.retries } : options.retries;
    options.stale = Math.max(options.stale || 0, 2000);
    options.update = options.update == null ? options.stale / 2 : options.update || 0;
    options.update = Math.max(Math.min(options.update, options.stale / 2), 1000);

    // Resolve to a canonical file path
    resolveCanonicalPath(file, options, (err, file) => {
        if (err) {
            return callback(err);
        }

        // Attempt to acquire the lock
        const operation = retry.operation(options.retries);

        operation.attempt(() => {
            acquireLock(file, options, (err, stat) => {
                if (operation.retry(err)) {
                    return;
                }

                if (err) {
                    return callback(operation.mainError());
                }

                // We now own the lock
                const lock = locks[file] = {
                    mtime: stat.mtime,
                    options,
                    lastUpdate: Date.now(),
                    mtimeChecker: createMtimeChecker(),
                };

                // We must keep the lock fresh to avoid staleness
                updateLock(file, options);

                callback(null, (releasedCallback) => {
                    if (lock.released) {
                        return releasedCallback &&
                            releasedCallback(Object.assign(new Error('Lock is already released'), { code: 'ERELEASED' }));
                    }

                    // Not necessary to use realpath twice when unlocking
                    unlock(file, { ...options, realpath: false }, releasedCallback);
                });
            });
        });
    });
}

function unlock(file, options, callback) {
    options = {
        fs,
        realpath: true,
        ...options,
    };

    // Resolve to a canonical file path
    resolveCanonicalPath(file, options, (err, file) => {
        if (err) {
            return callback(err);
        }

        // Skip if the lock is not acquired
        const lock = locks[file];

        if (!lock) {
            return callback(Object.assign(new Error('Lock is not acquired/owned by you'), { code: 'ENOTACQUIRED' }));
        }

        lock.updateTimeout && clearTimeout(lock.updateTimeout); // Cancel lock mtime update
        lock.released = true; // Signal the lock has been released
        delete locks[file]; // Delete from locks

        removeLock(file, options, callback);
    });
}

function check(file, options, callback) {
    options = {
        stale: 10000,
        realpath: true,
        fs,
        ...options,
    };

    options.stale = Math.max(options.stale || 0, 2000);

    // Resolve to a canonical file path
    resolveCanonicalPath(file, options, (err, file) => {
        if (err) {
            return callback(err);
        }

        // Check if lockfile exists
        options.fs.stat(getLockFile(file, options), (err, stat) => {
            if (err) {
                // If does not exist, file is not locked. Otherwise, callback with error
                return err.code === 'ENOENT' ? callback(null, false) : callback(err);
            }

            // Otherwise, check if lock is stale by analyzing the file mtime
            return callback(null, !isLockStale(stat, options));
        });
    });
}

function getLocks() {
    return locks;
}

function createMtimeChecker() {
    let precision;

    return (lockMtime, statMtime) => {
        // If lock time was not on the second we can determine precision
        if (!precision && lockMtime % 1000 !== 0) {
            precision = statMtime % 1000 === 0 ? 's' : 'ms';
        }

        if (precision === 's') {
            const lockTs = lockMtime.getTime();
            const statTs = statMtime.getTime();

            // Maybe the file system truncates or rounds...
            return Math.trunc(lockTs / 1000) === Math.trunc(statTs / 1000) ||
              Math.round(lockTs / 1000) === Math.round(statTs / 1000);
        }

        // Must be ms or lockMtime was on the second
        return lockMtime.getTime() === statMtime.getTime();
    };
}

// Remove acquired locks on exit
/* istanbul ignore next */
onExit(() => {
    for (const file in locks) {
        const options = locks[file].options;

        try { options.fs.rmdirSync(getLockFile(file, options)); } catch (e) { /* Empty */ }
    }
});

module.exports.lock = lock;
module.exports.unlock = unlock;
module.exports.check = check;
module.exports.getLocks = getLocks;
